### PR TITLE
sql/migrate: basic stmt scanning

### PR DESCRIPTION
**Pull request by a8m**
Don't review yet.

### CODE REVIEW

File: sql/migrate/lex.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"unicode/utf8"
+)
+
+// stmts provides a generic implementation for extracting
+// SQL statements from the given file string.
```
Comment (masseelch) [inline sql/migrate/lex.go:16]: What is meant by file string? The files contents? In that case name them like that ;-)

File: sql/migrate/lex.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+			return nil, fmt.Errorf("expect space after %q, got: %s", delimComment, input[:1])
+		}
+		input = strings.TrimSpace(input)
+		i := strings.Index(input, "\n")
+		if i == -1 {
+			return nil, errors.New("empty delimiter")
+		}
+		delim = strings.NewReplacer(`\n`, "\n", `\r`, "\r", `\t`, "\t").
+			Replace(input[:i])
+		input = strings.TrimSpace(input[i+1:])
```
Comment (masseelch) [inline sql/migrate/lex.go:66]: Add some comments here about what is happening please.

File: sql/migrate/lex.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	input string
+	pos   int    // current position.
+	width int    // size of latest rune.
+	start int    // start position of current statement.
+	depth int    // depth of parentheses.
+	delim string // configured delimiter.
+}
+
+const (
+	eos          = -1
```
Comment (masseelch) [inline sql/migrate/lex.go:46]: Is this a valid "not existing" rune?

File: sql/migrate/lex.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"unicode/utf8"
+)
+
+// stmts provides a generic implementation for extracting
+// SQL statements from the given file string.
+func stmts(input string) ([]string, error) {
```
Comment (masseelch) [inline sql/migrate/lex.go:17]: Can we change this to an `io.Reader`? This feels very un-goish.

File: sql/migrate/migrate.go [hunk 5]
```diff
@@ -778,7 +769,7 @@ func (f LocalFile) Name() string {
 
 // Read implements io.Reader.
 func (f LocalFile) Read(buf []byte) (int, error) {
-	return copy(buf, f.c.Bytes()), io.EOF
+	return f.c.Read(buf)
```
Comment (masseelch) [inline sql/migrate/migrate.go:772]: :100:

File: sql/migrate/testdata/lex/5_delimiter.sql [hunk 6]
```diff
@@ -0,0 +1,20 @@
+-- atlas:delimiter ATLAS_DELIMITER
```
Comment (masseelch) [inline sql/migrate/testdata/lex/5_delimiter.sql:1]: :wink:

File: sql/migrate/testdata/lex/5_delimiter.sql [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+DELIMITER $$
+
+CREATE DEFINER='boring' PROCEDURE proc ()
+    COMMENT 'ATLAS_DELIMITER'
```
Comment (masseelch) [inline sql/migrate/testdata/lex/5_delimiter.sql:6]: :wink:

**Review by masseelch — COMMENTED** [review]
I have one question (second comment).

File: sql/migrate/lex.go (hunk 2 above)
Comment (a8m) [reply sql/migrate/lex.go:66]: When scanning a delimiter string like `\n`, we get an escaped one. i.e. 2 bytes, `\\` and `n`.

File: sql/migrate/lex.go (hunk 4 above)
Comment (a8m) [reply sql/migrate/lex.go:18]: not really. I'm using the `strings` functions and Reader will make my work harder. Also, parsers usually work on bytes or strings.

File: sql/migrate/lex.go (hunk 2 above)
Comment (masseelch) [reply sql/migrate/lex.go:66]: I know. Still not easy to read and this is a good example of where a comment brings great value to a reviewer or when touching the code again.

Comment (a8m) [reply sql/migrate/lex.go:66]: Agree, I already updated the code before commenting here.

**Review by masseelch — APPROVED** [review]

**Pull request merged**